File: bench/suite/src/st/pidigits/cpp/pidigits.cpp
// pidigits, single-threaded C++ implementation using the GMP C API.
// Unbounded spigot algorithm (step-by-step) as described by the Computer
// Language Benchmarks Game.
#include <gmp.h>

#include <cstdio>
#include <cstdlib>
#include <string>

namespace {

class Spigot {
  public:
    Spigot() {
        mpz_init(tmp1_);
        mpz_init(tmp2_);
        mpz_init_set_ui(acc_, 0);
        mpz_init_set_ui(den_, 1);
        mpz_init_set_ui(num_, 1);
    }
    ~Spigot() {
        mpz_clear(tmp1_);
        mpz_clear(tmp2_);
        mpz_clear(acc_);
        mpz_clear(den_);
        mpz_clear(num_);
    }
    Spigot(const Spigot &) = delete;
    Spigot &operator=(const Spigot &) = delete;

    // Advances until the next digit is known and returns it.
    unsigned next_digit() {
        for (;;) {
            next_term(++k_);
            if (mpz_cmp(num_, acc_) > 0)
                continue;
            unsigned d = extract_digit(3);
            if (d != extract_digit(4))
                continue;
            eliminate_digit(d);
            return d;
        }
    }

  private:
    unsigned extract_digit(unsigned nth) {
        mpz_mul_ui(tmp1_, num_, nth);
        mpz_add(tmp2_, tmp1_, acc_);
        mpz_tdiv_q(tmp1_, tmp2_, den_);
        return static_cast<unsigned>(mpz_get_ui(tmp1_));
    }
    void eliminate_digit(unsigned d) {
        mpz_submul_ui(acc_, den_, d);
        mpz_mul_ui(acc_, acc_, 10);
        mpz_mul_ui(num_, num_, 10);
    }
    void next_term(unsigned k) {
        unsigned k2 = k * 2U + 1U;
        mpz_addmul_ui(acc_, num_, 2U);
        mpz_mul_ui(acc_, acc_, k2);
        mpz_mul_ui(den_, den_, k2);
        mpz_mul_ui(num_, num_, k);
    }

    mpz_t tmp1_, tmp2_, acc_, den_, num_;
    unsigned k_ = 0;
};

} // namespace

int main(int argc, char **argv) {
    const unsigned n = argc > 1 ? static_cast<unsigned>(std::atoi(argv[1])) : 27;
    Spigot spigot;
    std::string line;
    for (unsigned i = 1; i <= n; ++i) {
        line.push_back(static_cast<char>('0' + spigot.next_digit()));
        if (line.size() == 10) {
            std::printf("%s\t:%u\n", line.c_str(), i);
            line.clear();
        }
    }
    if (!line.empty()) {
        line.resize(10, ' ');
        std::printf("%s\t:%u\n", line.c_str(), n);
    }
}

File: bench/suite/src/st/pidigits/go/pidigits.go
// pidigits: single-threaded, same algorithm as the Benchmarks Game description
// (unbounded spigot, digit extraction by comparing (3*numer+accum)/denom and
// (4*numer+accum)/denom). Bignums via the standard library math/big.
package main

import (
	"bufio"
	"fmt"
	"math/big"
	"os"
	"strconv"
)

var (
	acc  = big.NewInt(0)
	den  = big.NewInt(1)
	num  = big.NewInt(1)
	tmp1 = new(big.Int)
	tmp2 = new(big.Int)
)

func extractDigit(nth int64) int64 {
	tmp1.Mul(num, big.NewInt(nth))
	tmp2.Add(tmp1, acc)
	tmp1.Quo(tmp2, den)
	return tmp1.Int64()
}

func nextTerm(k int64) {
	k2 := big.NewInt(k*2 + 1)
	tmp1.Lsh(num, 1)
	acc.Add(acc, tmp1)
	acc.Mul(acc, k2)
	den.Mul(den, k2)
	num.Mul(num, big.NewInt(k))
}

func eliminateDigit(d int64) {
	tmp1.Mul(den, big.NewInt(d))
	acc.Sub(acc, tmp1)
	acc.Mul(acc, big.NewInt(10))
	num.Mul(num, big.NewInt(10))
}

func main() {
	n := 27
	if len(os.Args) > 1 {
		n, _ = strconv.Atoi(os.Args[1])
	}
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	var line [10]byte
	i := 0
	for k := int64(1); i < n; k++ {
		nextTerm(k)
		if num.Cmp(acc) > 0 {
			continue
		}
		d := extractDigit(3)
		if d != extractDigit(4) {
			continue
		}
		line[i%10] = byte('0' + d)
		i++
		if i%10 == 0 {
			fmt.Fprintf(out, "%s\t:%d\n", line[:], i)
		}
		eliminateDigit(d)
	}
	if i%10 != 0 {
		for j := i % 10; j < 10; j++ {
			line[j] = ' '
		}
		fmt.Fprintf(out, "%s\t:%d\n", line[:], i)
	}
}
